ably/realtime/connectionmanager.py

```diff
@@ -166,13 +168,34 @@ def on_disconnected(self, msg: dict):
                     log.info("No fallback host to try for disconnected protocol message")
 
     async def on_error(self, msg: dict, exception: AblyException):
-        if msg.get('channel') is None:  # RTN15i
-            self.enact_state_change(ConnectionState.FAILED, exception)
-            if self.transport:
-                await self.transport.dispose()
-            raise exception
+        error = msg.get('error')
+        code = error.get('code')
+        # RTN14b
+        if is_token_error(code) and msg.get('channel') is None:
+            if isinstance(self.__error_reason, AblyException):
+                previous_error_code = self.__error_reason.code
+                if not is_token_error(previous_error_code):
+                    try:
+                        await self.ably.auth.authorize()
+                    except Exception as e:
+                        log.exception(f"Attempt to renew token fails: {e}")
+                        self.notify_state(ConnectionState.DISCONNECTED, e)
+                    return
+                self.notify_state(ConnectionState.DISCONNECTED, AblyException.from_dict(error))
```

Comment:
we should be using `self.__fail_state` here since you can get a token error when connecting from `SUSPENDED`